### DISCUSSION TITLE
Filter and search configuration of relationship tables

### DISCUSSION THREAD

**Discussion by ali-d-coded** [Power-Components/livewire-powergrid#327] [Q&A] [answered]
How can I configure the filter and search of a table having so many relations? I'm getting an error SQL saying value ambiguous. Nothing works, search, sort, nothing!!!
Please help me, guys. Anybody, please explain how it's done. It'll be so much helpful If I get an example of how it's done.

**Comment by dansysanalyst** [comment]
Hi @Ali-D-Coded,

Could you please provide us with some more details, so we can try to assist you better?

- What are your tables and relationships?
- What are you trying to accomplish?
- What is the result you get?

😊

**Reply by ali-d-coded** [reply]
`   <?php

namespace App\Http\Livewire;

use App\Models\DataCardOrder;
use Illuminate\Support\Carbon;
use Illuminate\Database\QueryException;
use Illuminate\Database\Eloquent\Builder;
use PowerComponents\LivewirePowerGrid\Button;
use PowerComponents\LivewirePowerGrid\Column;
use PowerComponents\LivewirePowerGrid\PowerGrid;
use PowerComponents\LivewirePowerGrid\PowerGridEloquent;
use PowerComponents\LivewirePowerGrid\PowerGridComponent;
use PowerComponents\LivewirePowerGrid\Traits\ActionButton;
use PowerComponents\LivewirePowerGrid\Rules\Rule;

final class DataCardsReport extends PowerGridComponent
{
    use ActionButton;

    //Messages informing success/error data is updated.
    public bool $showUpdateMessages = true;

    /*
    |--------------------------------------------------------------------------
    |  Features Setup
    |--------------------------------------------------------------------------
    | Setup Table's general features
    |
    */
    public function setUp(): void
    {
        $this->showCheckBox()
            ->showPerPage()
            ->showSearchInput()
            ->showExportOption('download', ['excel', 'csv']);
    }

    /*
    |--------------------------------------------------------------------------
    |  Datasource
    |--------------------------------------------------------------------------
    | Provides data to your Table using a Model or Collection
    |
    */

    /**
    * PowerGrid datasource.
    *
    * @return  \Illuminate\Database\Eloquent\Builder<\App\Models\User>|null
    */
    public function datasource(): ?Builder
    {
        return DataCardOrder::query()->join('users', 'users.id', '=', 'data_card_orders.user_id')
         ->join('data_card_order_plans', 'data_card_order_plans.order_id', '=', 'data_card_orders.id')
        ->join('data_card_order_cards', 'data_card_order_cards.order_plan_id', '=', 'data_card_order_plans.id')
        ->join('data_card_plans', 'data_card_plans.id', '=', 'data_card_order_plans.plan_id')
        ->select('data_card_orders.*',
        // 'data_card_orders.created_at as date',
         'users.name as user_name',
        //  'data_card_order_plans.quantity as qty',
         'data_card_plans.code as code',
         'data_card_plans.wsp as wsp',
         'data_card_plans.srp as srp')->distinct();
    }

    /*
    |--------------------------------------------------------------------------
    |  Relationship Search
    |--------------------------------------------------------------------------
    | Configure here relationships to be used by the Search and Table Filters.
    |
    */

    /**
     * Relationship search.
     *
     * @return array<string, array<int, string>>
     */
    public function relationSearch(): array
    {
        return []; // I want to know how to configure this part?? I could,nt find it in the documentation
    }

    /*
    |--------------------------------------------------------------------------
    |  Add Column
    |--------------------------------------------------------------------------
    | Make Datasource fields available to be used as columns.
    | You can pass a closure to transform/modify the data.
    |
    */
    public function addColumns(): ?PowerGridEloquent
    {
        // dd($this->datasource());

        return PowerGrid::eloquent()
            ->addColumn('id')
            ->addColumn('order_id')
            ->addColumn('user_name')
            ->addColumn('total_amount')
            // ->addColumn('qty')
            ->addColumn('code')
            ->addColumn('wsp')
            ->addColumn('srp')
            ->addColumn('date', function(DataCardOrder $model) {
                return Carbon::parse($model->created_at)->format('d/m/Y H:i:s');
            })
            ->addColumn('updated_at_formatted', function(DataCardOrder $model) {
                return Carbon::parse($model->updated_at)->format('d/m/Y H:i:s');
            });
    }

    /*
    |--------------------------------------------------------------------------
    |  Include Columns
    |--------------------------------------------------------------------------
    | Include the columns added columns, making them visible on the Table.
    | Each column can be configured with properties, filters, actions...
    |
    */

     /**
     * PowerGrid Columns.
     *
     * @return array<int, Column>
     */
    public function columns(): array
    {
        return [
            Column::add()
                ->title('ID')
                ->field('id')
                ->makeInputRange(),

            Column::add()
                ->title('Order Id')
                ->field('order_id')
                ->sortable()
                ->searchable()
                ->makeInputText(),

            Column::add()
                ->title('Retailer')
                ->field('user_name'),
                // ->makeInputRange(),

        

            Column::add()
                ->title('Plan Code')
                ->field('code')
                ->makeInputRange(),

            Column::add()
                ->title('Purchase price')
                ->field('wsp')
                ->makeInputRange(),

            Column::add()
                ->title('Sales Price')
                ->field('srp')
                ->makeInputRange(),

            Column::add()
                ->title('Total Amount')
                ->field('total_amount')
                ->sortable()
                ->searchable(),

            Column::add()
                ->title('Date')
                ->field('date', 'created_at')
                ->searchable()
                ->sortable()
                // ->makeInputDatePicker('created_at'),

          
        ];
    }

}
`


This is my powergrid class, I'm combining multiple database tables to form this datatable, so the search and sort or any filter option are'nt working!!!, whenever I try to do any sort or searching I get this error 

SQLSTATE[23000]: Integrity constraint violation: 1052 Column 'order_id' in where clause is ambiguous (SQL: select count(*) as aggregate from `data_card_orders` inner join `users` on `users`.`id` = `data_card_orders`.`user_id` inner join `data_card_order_plans` on `data_card_order_plans`.`order_id` = `data_card_orders`.`id` inner join `data_card_order_cards` on `data_card_order_cards`.`order_plan_id` = `data_card_order_plans`.`id` inner join `data_card_plans` on `data_card_plans`.`id` = 


how can I solve this?? please help me, I'm stuck.

**Comment by dansysanalyst** [comment, marked as answer]
Material on `relationSearch`:

📚 Documentation: https://livewire-powergrid.com/table/column-filters.html#filter-by-relationship

📖 Example: https://github.com/Power-Components/powergrid-demo/blob/main/app/Http/Livewire/DishesTable.php

**Reply by ali-d-coded** [reply]
Thanks!! I got what I wanted.

**Reply by BenOussama180** [reply]
@Ali-D-Coded i have same problem with created_at column, it says value of it ambiguos while date filtering , how did u fix it ? here is my component : 
`<?php

namespace App\Http\Livewire\Customers\Tables;

use App\Models\Order;
use Illuminate\Support\Carbon;
use Illuminate\Database\QueryException;
use Illuminate\Database\Eloquent\Builder;
use PowerComponents\LivewirePowerGrid\Rules\{Rule, RuleActions};
use PowerComponents\LivewirePowerGrid\Traits\ActionButton;
use PowerComponents\LivewirePowerGrid\{Button, Column, Exportable, Footer, Header, PowerGrid, PowerGridComponent, PowerGridEloquent};

final class CustomerProjectsTableV2 extends PowerGridComponent
{
    use ActionButton;

    /*
    |--------------------------------------------------------------------------
    |  Features Setup
    |--------------------------------------------------------------------------
    | Setup Table's general features
    |
    */
    public function setUp(): array
    {
        return [
            Header::make()->showSearchInput(),
            Footer::make()
                ->showPerPage(50, [10, 20, 50, 100])
                ->showRecordCount(),
        ];
    }

    /*
    |--------------------------------------------------------------------------
    |  Datasource
    |--------------------------------------------------------------------------
    | Provides data to your Table using a Model or Collection
    |
    */

    /**
     * PowerGrid datasource.
     *
     * @return Builder<\App\Models\Order>
     */
    public function datasource(): Builder
    {
        return Order::query()
            ->leftJoin('projectables', 'projectables.projectable_id', '=', 'orders.id')
            ->leftJoin('customer_projects', 'customer_projects.id', '=', 'projectables.project_id')
            ->where('projectables.projectable_type', Order::class)
            ->where('customer_projects.customer_id', auth()->id())
            ->with('projects')
            ->select([
                'orders.id',
                'orders.created_at as created_at_orders',
                'customer_projects.project_name as projectName',
            ])
            ->groupBy('customer_projects.project_name');
    }

    /*
    |--------------------------------------------------------------------------
    |  Relationship Search
    |--------------------------------------------------------------------------
    | Configure here relationships to be used by the Search and Table Filters.
    |
    */

    /**
     * Relationship search.
     *
     * @return array<string, array<int, string>>
     */
    public function relationSearch(): array
    {
        return [];
    }

    /*
    |--------------------------------------------------------------------------
    |  Add Column
    |--------------------------------------------------------------------------
    | Make Datasource fields available to be used as columns.
    | You can pass a closure to transform/modify the data.
    |
    | ❗ IMPORTANT: When using closures, you must escape any value coming from
    |    the database using the `e()` Laravel Helper function.
    |
    */
    public function addColumns(): PowerGridEloquent
    {
        return PowerGrid::eloquent()
            ->addColumn('id')
            ->addColumn('projectName')
            ->addColumn('name_lower', fn (Order $model) => strtolower(e($model->name)))
            ->addColumn('created_at_orders')
            ->addColumn('created_at_formatted', fn (Order $model) => Carbon::parse($model->created_at_orders)->format('d/m/Y H:i:s'));
    }

    /*
    |--------------------------------------------------------------------------
    |  Include Columns
    |--------------------------------------------------------------------------
    | Include the columns added columns, making them visible on the Table.
    | Each column can be configured with properties, filters, actions...
    |
    */

    /**
     * PowerGrid Columns.
     *
     * @return array<int, Column>
     */
    public function columns(): array
    {
        return [
            Column::make('ID', 'id')
                ->searchable()
                ->sortable(),

            Column::make('Name', 'projectName')
                ->searchable()
                ->makeInputText('customer_projects.project_name')
                ->sortable(),

            Column::make('Created at', 'created_at_formatted', 'created_at_orders')
                ->makeInputDatePicker()
                ->searchable()
        ];
    }

    /*
    |--------------------------------------------------------------------------
    | Actions Method
    |--------------------------------------------------------------------------
    | Enable the method below only if the Routes below are defined in your app.
    |
    */

    /**
     * PowerGrid Order Action Buttons.
     *
     * @return array<int, Button>
     */

    /*
    public function actions(): array
    {
       return [
           Button::make('edit', 'Edit')
               ->class('bg-indigo-500 cursor-pointer text-white px-3 py-2.5 m-1 rounded text-sm')
               ->route('order.edit', ['order' => 'id']),

           Button::make('destroy', 'Delete')
               ->class('bg-red-500 cursor-pointer text-white px-3 py-2 m-1 rounded text-sm')
               ->route('order.destroy', ['order' => 'id'])
               ->method('delete')
        ];
    }
    */

    /*
    |--------------------------------------------------------------------------
    | Actions Rules
    |--------------------------------------------------------------------------
    | Enable the method below to configure Rules for your Table and Action Buttons.
    |
    */

    /**
     * PowerGrid Order Action Rules.
     *
     * @return array<int, RuleActions>
     */

    /*
    public function actionRules(): array
    {
       return [

           //Hide button edit for ID 1
            Rule::button('edit')
                ->when(fn($order) => $order->id === 1)
                ->hide(),
        ];
    }
    */
}
`

**Reply by luanfreitasdev** [reply]
Hello @BenOussama180 !, can you share the error / sql ?

**Reply by BenOussama180** [reply]
for thhat code above, i tried to make an alias for the created_at column inside orders table but it says  : 
![image](https://user-images.githubusercontent.com/81111919/220936340-41bf79f1-4bf5-4379-997a-e423891d34aa.png)

, if i remove the alias, it just says that the created_at col is ambiguous! @luanfreitasdev

**Reply by BenOussama180** [reply]
that error happens while using date filter!

**Reply by luanfreitasdev** [reply]
you select the dates and have this error?. Do the other filters work?

**Reply by BenOussama180** [reply]
yes other filters work because no other field is ambigus, but i think powergrid doesnt know abt aliasses and just treat them as regular columns in database, even tho in another table when i used the same alias techhnique it worked perfectly im lost xd

**Reply by luanfreitasdev** [reply]
Did you try like this?
```php
Column::make('Created at', 'created_at_formatted', 'orders.created_at')
            ->makeInputDatePicker('orders.created_at')
            ->searchable()
```

**Reply by BenOussama180** [reply]
yes tried that, it just says unknown column orders.created_at

**Reply by luanfreitasdev** [reply]
And this ?

```php
Column::make('Created at', 'created_at_formatted', 'created_at_orders')
            ->makeInputDatePicker('orders.created_at')
            ->searchable()
```

**Reply by BenOussama180** [reply]
i dont think i did, but i believe it'll throw the same error since the datefilter will filter with orders. created at , but ill give it a try and get back to u

**Reply by luanfreitasdev** [reply]
I need to debug this specific case more because it is join. Can you create a repository with only this information?

I'm working with dates in v4 and it would be nice if this was understood 😃. Thank you anyway

**Reply by BenOussama180** [reply]
sure i will create a smiliar case repository for this tomorrow and leave the link in this discussion

**Reply by BenOussama180** [reply]
```

Column::make('Created at', 'created_at_formatted', 'created_at_orders')
            ->makeInputDatePicker('orders.created_at')
            ->searchable()
 ```

i did try this it threw the same error :! SQLSTATE[42S22]: Column not found: 1054 Unknown column 'orders.created_at' in 'where clause'
the thing is it does filter out the table but it also throws the error haha its weird

**Reply by luanfreitasdev** [reply]
Does it work and throws errors at the same time?

**Reply by BenOussama180** [reply]
yes exactly when i remove the livewire popup error i can see the table has been correctly filtered

**Reply by luanfreitasdev** [reply]
Strange. Can you record a video and share the full flare please?

**Reply by BenOussama180** [reply]
here is the video , i uploaded it on https://streamable.com/ : 
Video Link : https://streamable.com/umlqll

**Reply by BenOussama180** [reply]
and here is a gif just incase something weird happens with streamble.com 
![ezgif-4-be4a984bf9](https://user-images.githubusercontent.com/81111919/221328335-b780341b-d396-4374-9fbf-23fd7680948b.gif)

**Reply by luanfreitasdev** [reply]
I will try to reproduce this error

**Reply by BenOussama180** [reply]
alright well i have customer_projects table which have project_name & customer_id columns
orders table which morphToMany(Project::class, 'projectable');
projectables table which is a polymorphic table ,

**Reply by BenOussama180** [reply]
you would say why would i even put Order model in my datasrouce if  i only want too display project names & their data, why not display directly Project model data , the thing is while displaying Project model data i want to filter the date with the orders inside it, if  orders were created in the range of created_at filter i want to display the project or no , and since we cant pass a callback to the filter date method to define and modify the behavior of it , i was forced to go with Order model and do some weird joins to obtain my data that i couldve gotten easilly with using project model!

**Reply by luanfreitasdev** [reply]
it seems a little complex to assemble this scenario for this specific case. Can you share a structure close to this?

**Reply by BenOussama180** [reply]
u knoow what, let me create a repo for this specific case and i'll put the link here give me 10min

**Reply by luanfreitasdev** [reply]
I believe this PR can also help https://github.com/Power-Components/livewire-powergrid/pull/864

**Reply by BenOussama180** [reply]
surprisignly , it works fine in the demo project xDD 
repo link : https://github.com/BenOussama180/poweregrid-test-case

i got no idea why its not working on my project, mybe i should update poowergrid tolatest version i am using 3.5.13

**Reply by luanfreitasdev** [reply]
Hello @BenOussama180 . I downloaded your project and it's really working

**Reply by BenOussama180** [reply]
yea i still cant figure out why its not working on my project , ill try updating powergrid maybe it'll fix
